system-contracts/test/L1Messenger.spec.ts

```diff
@@ -0,0 +1,526 @@
+      const emptyKV = ethers.utils.hexlify([]);
+      await expect(l1Messenger.connect(l1MessengerAccount).sendL2ToL1Log(isService, emptyKV, emptyKV)).to.be.rejected;
+    });
+
+    it("should revert when called by the system contract with key & value > 32 bytes", async () => {
+      const oversizedKV = ethers.utils.hexlify(randomBytes(33));
+      await expect(l1Messenger.connect(l1MessengerAccount).sendL2ToL1Log(isService, oversizedKV, oversizedKV)).to.be
+        .rejected;
+    });
+
+    it("should revert when called by the system contract with key & value < 32 bytes", async () => {
+      const undersizedKV = ethers.utils.hexlify(randomBytes(31));
+      await expect(l1Messenger.connect(l1MessengerAccount).sendL2ToL1Log(isService, undersizedKV, undersizedKV)).to.be
+        .rejected;
+    });
+  });
+
+  describe("sendToL1", async () => {
+    it("should emit L1MessageSent & L2ToL1LogSent events", async () => {
+      const expectedKey = ethers.utils
+        .hexZeroPad(ethers.utils.hexStripZeros(l1MessengerAccount.address), 32)
+        .toLowerCase();
+      await expect(l1Messenger.connect(l1MessengerAccount).sendToL1(message))
+        .to.emit(l1Messenger, "L1MessageSent")
+        .withArgs(l1MessengerAccount.address, messageHash, message)
+        .and.to.emit(l1Messenger, "L2ToL1LogSent")
+        .withArgs([0, true, txNumberInBlock, l1Messenger.address, expectedKey, messageHash]);
+    });
+  });
+
+  describe("requestBytecodeL1Publication", async () => {
+    it("should revert when not called by known code storage contract", async () => {
+      const byteCodeHash = ethers.utils.hexlify(randomBytes(32));
+      await expect(l1Messenger.requestBytecodeL1Publication(byteCodeHash)).to.be.rejectedWith("Inappropriate caller");
+    });
+
+    it("shoud revert when byteCodeHash < 32 bytes, called by known code system contract", async () => {
+      const byteCodeHash = ethers.utils.hexlify(randomBytes(8));
+      await expect(l1Messenger.connect(knownCodeStorageAccount).requestBytecodeL1Publication(byteCodeHash)).to.be
+        .rejected;
+    });
+
+    it("shoud revert when byteCodeHash > 32 bytes, called by known code system contract", async () => {
+      const byteCodeHash = ethers.utils.hexlify(randomBytes(64));
+      await expect(l1Messenger.connect(knownCodeStorageAccount).requestBytecodeL1Publication(byteCodeHash)).to.be
+        .rejected;
+    });
+
+    it("shoud emit event, called by known code system contract", async () => {
+      await expect(
+        l1Messenger.connect(knownCodeStorageAccount).requestBytecodeL1Publication(bytecodeHash, { gasLimit: 130000000 })
+      )
+        .to.emit(l1Messenger, "BytecodeL1PublicationRequested")
+        .withArgs(bytecodeHash);
+    });
+  });
+});
+
+// helpers
+interface DataPair {
+  lengthBytes: string;
+  content: string;
+}
+
+function createTotalL2ToL1PubdataAndStateDiffs(
```

Comment:
This is a good step forward, but still, lots of the same variables are almost always supplied to `createTotalL2ToL1PubdataAndStateDiffs`.

I understand that the input contains lots of params & so I would probably suggest considering something like:

```
class L1MessengerPubdataEmulator {
   // 
   // ... Here there are methods to append a log, a message, etc.
   // ... This methods could also invoke the l1messenger itself if needed
   // 

   buildTotalL2ToL1PubdataAndStateDiffs(overrides: { ... }) {
   //  Here the pubdata is built with correct parts, except for overrides that allow to supply non-correct values.
   }
}
```